File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {
    private void graphDFSIve(int currVertex, Map<Integer, List<Integer>> adjList, Map<Integer, Boolean> visitedVerticesMap){
        Stack<Integer> verticesStack = new Stack<>();
        verticesStack.push(currVertex);
        visitedVerticesMap.put(currVertex, true);

        while(!verticesStack.isEmpty()){
            currVertex = verticesStack.pop();

            List<Integer> currVertexNeighbors = adjList.get(currVertex);

            for(int idxI = 0; idxI < currVertexNeighbors.size(); idxI++){
                int neighborVertex = currVertexNeighbors.get(idxI);

                if(!visitedVerticesMap.containsKey(neighborVertex)){
                    visitedVerticesMap.put(neighborVertex, true);
                    verticesStack.push(neighborVertex);
                }
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int mRows = isConnected.length;
        int nCols = isConnected[0].length;

        Map<Integer, List<Integer>> adjList = new HashMap<>();

        for(int i = 0; i < mRows; i++){
            for(int j = 0; j < nCols; j++){
                int currCellVal = isConnected[i][j];

                if(i == j || currCellVal == 0){
                    continue;
                }

                int cityVertex1 = i;
                int cityVertex2 = j;

                if(!adjList.containsKey(cityVertex1)){
                    adjList.put(cityVertex1, new ArrayList<>());
                }

                adjList.get(cityVertex1).add(cityVertex2);
            }
        }

        Map<Integer, Boolean> visitedVerticesMap = new HashMap<>();
        int totalNumOfProvinces = 0;
        
        for(int currCityVertex = 0; currCityVertex < mRows; currCityVertex++){
            if(!adjList.containsKey(currCityVertex) && !visitedVerticesMap.containsKey(currCityVertex)){
                totalNumOfProvinces += 1;
            }
            else if(!visitedVerticesMap.containsKey(currCityVertex)){
                totalNumOfProvinces += 1;
                graphDFSIve(currCityVertex, adjList, visitedVerticesMap);
            }
        }
        
        return totalNumOfProvinces;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        # Solved during DSA Session 12 on June 14
        m_rows = len(isConnected)
        n_cols = len(isConnected[0])

        adjacency_list = {}
        visited_vertices_map = [False] * m_rows

        total_num_of_provinces = 0

        for row_idx in range(m_rows):
            for col_idx in range(n_cols):
                curr_cell_val = isConnected[row_idx][col_idx]

                if row_idx == col_idx or curr_cell_val == 0:
                    continue
                
                city_vertex1 = row_idx
                city_vertex2 = col_idx

                if city_vertex1 not in adjacency_list:
                    adjacency_list[city_vertex1] = []
                
                adjacency_list[city_vertex1].append(city_vertex2)
        
        # print(adjacency_list)

        def graphDFSIve(curr_city_vertex):
            vertices_stack = []

            vertices_stack.append(curr_city_vertex)

            while len(vertices_stack) > 0:
                curr_city_vertex = vertices_stack.pop()

                if not visited_vertices_map[curr_city_vertex]:
                    visited_vertices_map[curr_city_vertex] = True

                    curr_city_vertices_neighbors = adjacency_list[curr_city_vertex]
                    for curr_city_vertex_neighbor in curr_city_vertices_neighbors:
                        if not visited_vertices_map[curr_city_vertex_neighbor]:          
                            vertices_stack.append(curr_city_vertex_neighbor)
        
        for curr_city_vertex in range(m_rows):
            if curr_city_vertex not in adjacency_list and not visited_vertices_map[curr_city_vertex]:
                total_num_of_provinces += 1
            elif not visited_vertices_map[curr_city_vertex]:
                total_num_of_provinces += 1
                graphDFSIve(curr_city_vertex)
        
        return total_num_of_provinces



